### ISSUE TITLE
Alternative access order for the same buffer can bring big perf win

### ISSUE THREAD

**Issue by shunting314** [pytorch/pytorch#126913]
### 🐛 Describe the bug

Check this softmax kernel generated by inductor: https://gist.github.com/shunting314/16bf79d906bd2e929a62c0b2f3c02150  (call it k1)

If we reverse the access order for the second for loop from:
```
for roffset in range(0, rnumel, RBLOCK):
```
to 
```
for roffset in range(rnumel - rnumel % RBLOCK, -RBLOCK, -RBLOCK):
```
we get k2 (https://gist.github.com/shunting314/e749c3766757adaed729b51d38cd3169 )

k2 is 1.54 x faster than k1 (5.231ms v.s. 8.067ms). The speedup is mainly due to more cache hit.

Credit to llm.c since I learn the idea from there.  This is probably something we can apply in general in inductor.

### Versions

.

**Comment by jansel** [comment]
Interesting, maybe we should just always alternate the order for multiple reduction loops.

**Comment by trueoneplusone** [comment]
AI assistance disclosure: ChatGPT helped inspect the current Triton codegen path and set up the experiments below.

I'd like to work on this issue. I re-ran the core locality idea on current Triton 3.7.1 / RTX 5090 using a modern three-pass softmax-style kernel: forward-only traversal vs forward→reverse→forward. Negative-step `tl.range` works on the current stack and the outputs are numerically equivalent. Across `RBLOCK={128,256,512,1024,2048}`, the alternating traversal is consistently about **5.3–5.5% faster** on this Blackwell GPU (roughly 2.86 ms → 2.71 ms at representative sizes), so the effect still exists and is not a single block-size artifact.

My plan is to implement this conservatively in Inductor codegen rather than add an autotune dimension: alternate safe sequential reduction-loop segments, initially excluding cooperative/nested cases and any path that requires loop-carried pointer advancement. I’ll validate the *actual torch.compile-generated kernel* (source shape + correctness + latency), not just the standalone Triton microbenchmark, before sending a PR.
